### ISSUE TITLE
Can't see Storm in my graphite UI

### ISSUE THREAD

**Issue by alecinvan** [staslev/storm-metrics-reporter#1]
Hi, 

I like to report Storm's built-in metrics using the storm-metrics-reporter, here how I make it, first
1. add dependency in my topology pom
   <dependency>
   <groupId>com.github.staslev</groupId>
   <artifactId>storm-metrics-reporter</artifactId>
   <version>1.0</version>
   </dependency>
2. add this snippet my code:

import com.github.staslev.storm.metrics.graphite.SimpleGraphiteStormMetricGauge;
import com.github.staslev.storm.metrics.MetricReporterConfig;
import com.github.staslev.storm.metrics.MetricReporter;

conf.put(MetricReporter.METRICS_HOST, "10.100.70.29");
conf.put(MetricReporter.METRICS_PORT, "2003");
conf.registerMetricsConsumer(MetricReporter.class,
                               new MetricReporterConfig("(.execute.|.latency.|.capacity.)",
                                                        SimpleGraphiteStormMetricGauge.class.getCanonicalName()),
                               1);
I successfully build the jar and submit the topology.
1. Now I assume my topology has registered the storm-metrics-reporter and sending buildin metrics to graphite. 
2. This is how I start graphite-web 
   PYTHONPATH=`pwd`/storage/whisper ./bin/run-graphite-devel-server.py --port=8085 --libs=`pwd`/webapp /opt/graphite 1>/opt/graphite/storage/log/webapp/process.log 2>&1 &
3. carbon.conf
   [cache]
   LINE_RECEIVER_INTERFACE = 0.0.0.0
   LINE_RECEIVER_PORT = 2003
   CACHE_QUERY_INTERFACE = 0.0.0.0
   CACHE_QUERY_PORT = 7002

However, I can't see storm metrics in graphite, any ideas?

thanks

AL

**Comment by staslev** [comment]
Hi AL,

Edit:
Did you try inspecting the worker's log for errors?

Let's assume your Graphite setup is operational for now and verify a couple of Storm related things:
1.  The metrics bolt is up and running.
2.  `SimpleGraphiteStormMetricGauge` is indeed sending metrics out to your Graphite.

There are basically 2 ways you can approach this.

### The check-it-in-test way:

If you could run your topology in a local cluster mode (embedded) from within your IDE (very useful for tests) you could place a break point in the `process` method of `SimpleGraphiteStormMetricGauge` and see if you're getting there, and essentially inspect if anything goes wrong during the send process. 

### The check-it-in-production way:

To verify (1), could you please take a look at Storm's UI, locate your topology, click on it and then click the `Show System Stats` button at the bottom of the page. Then look for `__metricscom.github.staslev.storm.metrics.MetricReporter` in the `Bolts (All time)` section. Then click it, and locate the machine it's running on by inspecting the `Executors` section.

To verify (2), could you please go to that machine and run the following command:
`tcpdump -i any -A -s0 port 20003`, you should be able to see your metrics sent out about every minute or so (that's the default and I see you haven't changed it).

Could you try (any of) the above and let me know how it goes?

-Stas

**Comment by alecinvan** [comment]
Definitely try that, one thing I need to clarify, my carbon-cache:LINE_RECEIVER_PORT is set as 2003, can I set MetricReporter.METRICS_PORT =2003?

thanks
AL

**Comment by alecinvan** [comment]
I just tried in The check-it-in-production way, I did see __metricscom.github.staslev.storm.metrics.MetricReporter in the Bolts (All time) section. Click in, it shows: 
Executors
[6-6]}  4m 48s  pof-kstorm-dev2     6703    40  0   0.001   0.800   300     0.733   300     0
Where pof-kstorm-dev2 is the node running, everything looks not bad so far. I got into pof-kstorm-dev2, run
root@pof-kstorm-dev2:/opt/graphite/conf# tcpdump -i any -A -s0 port 2003
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 65535 bytes
.
then stuck here, seems SimpleGraphiteStormMetricGauge is not sending anyting, btw, I didn't use storm-metrics-reporter-1.1, but 1.0, since with 1.1, I am having trouble to build the jar. It says can't find SimpleGraphiteStormMetricGauge class, I shift to 1.0, it worked. 

thanks

**Comment by staslev** [comment]
I think 2003 should be fine.

Edit: Just went through the code (and wiki) again, and the problem may be a whole lot simpler. Could you try changing: `(.execute.|.latency.|.capacity.)` to `.*` ? I think the regular expression example in the wiki may be faulty :(

`SimpleGraphiteStormMetricGauge` was renamed to `SimpleGraphiteStormMetricProcessor`, you can go ahead and try `1.2` which was published today, wiki was updated as well.
-  Did you wait a while for the metrics to be emitted? Metrics are sent once in a minute or so. You can expedite things by setting `config.put(Config.TOPOLOGY_BUILTIN_METRICS_BUCKET_SIZE_SECS, 10)` and have Storm emit metrics every 10 seconds (or any other duration).
-  Did you get a chance to look at your topology worker's logs (the one that runs the metrics bolt)? In case sending to Graphite fails there should be a log printout, "Unable to send metric %s to Graphite server %s:%d".

Basically the goal now is to determine if the metrics make it `SimpleGraphiteStormMetricProcessor` and just fail to be sent to Graphite, or, if they fail to make it to `SimpleGraphiteStormMetricProcessor` in the first place.

**Comment by staslev** [comment]
Damn, the markdown made the wildcard in the regexp go away, so `(.*execute.*|.*latency.*|.*capacity.*)` showed up as `(.execute.|.latency.|.capacity.)`. Just fixed the wiki, let me know how it goes with the "new" regexp.

**Comment by alecinvan** [comment]
Oops, I pull out the 1.2 version, building with no error, change code as:

conf.put(MetricReporter.METRICS_HOST, "10.100.70.29");
conf.put(MetricReporter.METRICS_PORT, "2003");
conf.put(Config.TOPOLOGY_BUILTIN_METRICS_BUCKET_SIZE_SECS, 10);
conf.registerMetricsConsumer(MetricReporter.class,
                                         new MetricReporterConfig("(._execute._|._latency._|._capacity._)", SimpleGraphiteStormMetricProcessor.class.getCanonicalName()),
                                         1);

I don't see any error in storm UI, but when I tail -f worker-6703.log, I am getting such errors:

2015-03-12T10:27:49.162-0700 c.g.s.s.m.y.SimpleGraphiteStormMetricProcessor [ERROR] Unable to send metric "pof-kstorm-dev2.6703.b-1":type="8",name="execute_count" to Graphite server 10.100.70.29:2003
java.lang.NullPointerException: null
        at com.yammer.metrics.reporting.GraphiteReporter.sendToGraphite(GraphiteReporter.java:271) ~[stormjar.jar:na]
        at com.yammer.metrics.reporting.GraphiteReporter.sendObjToGraphite(GraphiteReporter.java:265) ~[stormjar.jar:na]
        at com.yammer.metrics.reporting.GraphiteReporter.processGauge(GraphiteReporter.java:304) ~[stormjar.jar:na]
        at com.github.staslev.storm.metrics.yammer.SimpleGraphiteStormMetricProcessor.process(SimpleGraphiteStormMetricProcessor.java:66) ~[stormjar.jar:na]
        at com.github.staslev.storm.metrics.MetricReporter.handleDataPoints(MetricReporter.java:104) [stormjar.jar:na]
        at backtype.storm.metric.MetricsConsumerBolt.execute(MetricsConsumerBolt.java:55) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.daemon.executor$fn__3441$tuple_action_fn__3443.invoke(executor.clj:633) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.daemon.executor$mk_task_receiver$fn__3364.invoke(executor.clj:401) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.disruptor$clojure_handler$reify__1447.onEvent(disruptor.clj:58) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:125) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:99) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:80) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.daemon.executor$fn__3441$fn__3453$fn__3500.invoke(executor.clj:748) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.util$async_loop$fn__464.invoke(util.clj:463) [storm-core-0.9.3.jar:0.9.3]
        at clojure.lang.AFn.run(AFn.java:24) [clojure-1.5.1.jar:na]
        at java.lang.Thread.run(Thread.java:745) [na:1.7.0_65]

2015-03-12T10:27:34.173-0700 c.g.s.s.m.y.SimpleGraphiteStormMetricProcessor [ERROR] Unable to send metric "pof-kstorm-dev2.6703.system":type="-1",name="transfer.capacity" to Graphite server 10.100.70.29:2003
java.lang.NullPointerException: null
        at com.yammer.metrics.reporting.GraphiteReporter.sendToGraphite(GraphiteReporter.java:271) ~[stormjar.jar:na]
        at com.yammer.metrics.reporting.GraphiteReporter.sendObjToGraphite(GraphiteReporter.java:265) ~[stormjar.jar:na]
        at com.yammer.metrics.reporting.GraphiteReporter.processGauge(GraphiteReporter.java:304) ~[stormjar.jar:na]
        at com.github.staslev.storm.metrics.yammer.SimpleGraphiteStormMetricProcessor.process(SimpleGraphiteStormMetricProcessor.java:66) ~[stormjar.jar:na]
        at com.github.staslev.storm.metrics.MetricReporter.handleDataPoints(MetricReporter.java:104) [stormjar.jar:na]
        at backtype.storm.metric.MetricsConsumerBolt.execute(MetricsConsumerBolt.java:55) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.daemon.executor$fn__3441$tuple_action_fn__3443.invoke(executor.clj:633) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.daemon.executor$mk_task_receiver$fn__3364.invoke(executor.clj:401) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.disruptor$clojure_handler$reify__1447.onEvent(disruptor.clj:58) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:125) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:99) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:80) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.daemon.executor$fn__3441$fn__3453$fn__3500.invoke(executor.clj:748) [storm-core-0.9.3.jar:0.9.3]
        at backtype.storm.util$async_loop$fn__464.invoke(util.clj:463) [storm-core-0.9.3.jar:0.9.3]
        at clojure.lang.AFn.run(AFn.java:24) [clojure-1.5.1.jar:na]
        at java.lang.Thread.run(Thread.java:745) [na:1.7.0_65]

Seems that I could not connect to graphite server? actually graphite server is on pof-kstorm-dev2, same node with the machine running metricReporter bolt. 

thanks

**Comment by staslev** [comment]
If you try `telent 10.100.70.29 2003`, do you get any errors?
Are you sure your Graphite deployment is operational?

P.S.
In your comment you still specify the old regexp, the one with no wildcards. Is that a typo?

**Comment by staslev** [comment]
I think I see the problem, might be my bad.

`SimpleGraphiteStormMetricProcessor.java` line `66`:

``` Java
graphiteReporter.processGauge(metricName, gauge, taskInfo.timestamp);
```

Might need to be:

``` Java
graphiteReporter.start(30, TimeUnit.MILLISECONDS); // start the Graphite reporter (!)
graphiteReporter.processGauge(metricName, gauge, taskInfo.timestamp);
```

Can you try this change and let me know if it helps?
Are you building the storm-metrics-reporter yourself, or using the maven repository to get the jar?

If it does, I'll fix it and release a new version.

**Comment by alecinvan** [comment]
telnet 10.100.70.29 2003
Trying 10.100.70.29...
Connected to 10.100.70.29.
Escape character is '^]'.

Currently I am kinda worry if my Graphite is truly workable, my os is Ubuntu, I followed this page: 
http://www.infoq.com/articles/graphite-intro to get graphite up, seems I am able to see CPUUsage ... in graphite UI. As matter of fact, I found many tutorial of graphite deployment won't make me install graphite smoothly, :(.

**Comment by staslev** [comment]
The fact you have managed to connect only makes me think it is my bad even more.
Anyhow, if you could try the change above and see if it helps, we might be able to get things working for you :)

**Comment by alecinvan** [comment]
Thank you very much, staslev, did you push to maven? I just need to rebuild the topo jar?

**Comment by staslev** [comment]
I think I've found the problem, I'm not using the GraphiteReporter properly (we use a different one in our PROD, so this particular piece of code has not been tested properly).

I'll figure it out and push a new version to maven. Will let you know once it's out, I hope it will be pretty soon.

**Comment by alecinvan** [comment]
Great! thank you very much, meanwhile, I may re-deploy graphite on ubuntu. :)

**Comment by staslev** [comment]
Please try this one and let me know if it's better.

``` XML
<dependency>
  <groupId>com.github.staslev</groupId>
  <artifactId>storm-metrics-reporter</artifactId>
  <version>1.4-SNAPSHOT</version>
</dependency>
```

**Comment by alecinvan** [comment]
Having this error, 
[ERROR] Failed to execute goal on project kafka-storm-ingress-writer: Could not resolve dependencies for project storm.ingress:kafka-storm-ingress-writer:jar:0.0.1-SNAPSHOT: Failure to find com.github.staslev:storm-metrics-reporter:jar:1.4-SNAPSHOT in http://clojars.org/repo was cached in the local repository, resolution will not be reattempted until the update interval of clojars.org has elapsed or updates are forced -> [Help 1]

when I am building the jar.

**Comment by staslev** [comment]
Please add this to your pom to allow a SNAPSHOT version to be downloaded.

``` XML
<repositories>
    <!-- Added to get the storm-metrics-reporter-1.4-SNAPSHOT, can be removed when it's released -->
    <repository>
      <id>oss.sonatype.org-snapshot</id>
      <url>http://oss.sonatype.org/content/repositories/snapshots</url>
      <releases>
        <enabled>false</enabled>
      </releases>
      <snapshots>
        <enabled>true</enabled>
      </snapshots>
    </repository>
</repositories>
```

**Comment by alecinvan** [comment]
It is awesome, it works, and I can see it on graphite UI, one more question, my storm cluster has 3 nodes, I see pof-kstorm-dev2 is the machine to execute (worker 6703) in storm UI, but in graphite, I can see both  pof-kstorm-dev1 and  pof-kstorm-dev2. Structure like this:

-Storm
     |- pipe-test-graphite (my topology name)
          |- *
          |- jvm
          |- pof-kstorm-dev1
                    |- 6703 (worker)
                             |- *
                             |- $mastercoord-bg0
                             |- acker
                             |- b-0
                             |- spout0
                             |- system  
          |- pof-kstorm-dev2
                    |- 6703  
                             |- *
                             |- $spoutcoord-spout0
                             |- acker
                             |- b-1
                             |- system  

 Can you explain a bit of this ? thanks a lot for your help, that is a great tool.

**Comment by staslev** [comment]
Great I'll make it an official version then.

I'm not sure I get the structure right, but my guess is that your topology has 2 workers, both on port 6703, but one is being run from a node named "pof-kstorm-dev1" and the other "pof-kstorm-dev1".

Since metrics for a given topology come from all workers involved in running it, your're seeing both nodes.

If you take a look at the javadoc for `SimpleGraphiteStormMetricProcessor`:

```
   -WorkerHost (i.e., pof-kstorm-dev1, pof-kstorm-dev1)
      -WorkerPort (i.e., 6703)
        -ComponentName (i.e., acker, spout0, etc)
          -TaskId (not mentioned in your comment)
            -OperationName (not mentioned in your comment)
              -value  (not mentioned in your comment)
```

Makes sense?

In general, the way one prefers to see his metrics tree is a matter of taste, the tree suggested (or rather, implemented) by `SimpleGraphiteStormMetricProcessor` is just an example. By deriving from `StormMetricProcessor` one can configure a different "output method", so to speak.

**Comment by alecinvan** [comment]
Awesome, thank you very much. :)   Will dive into this page: https://github.com/staslev/storm-metrics-reporter/wiki/Custom-metrics-%26-Yammer-metrics-API

**Comment by staslev** [comment]
Sure thing :)

I'll push an official version to maven tomorrow, and close this issue. 
Feel free to drop by when you have to report custom metrics :)

**Comment by staslev** [comment]
A new version with the fix is out (see below), closing the issue.

``` XML
<dependency>
  <groupId>com.github.staslev</groupId>
  <artifactId>storm-metrics-reporter</artifactId>
  <version>1.4.1</version>
</dependency>
```
